### ISSUE TITLE
UnboundLocalError when using pytest-splinter in Docker container

### ISSUE THREAD

**Issue by jmt-transloc** [cobrateam/splinter#755]
I am attempting to shift my automation at work from Selenium to Splinter. As part of this, I have chosen to utilize the pytest-selenium plugin for better QOL. Unfortunately, it appears that Splinter is raising an UnboundLocalError when run from a Docker container, not when running from command line. This error is raised at line 68 in browser.py within the Splinter library.

Any thoughts on what could be happening here? 

```
splinter_webdriver = 'chrome', retry_count = 3

    def get_browser(splinter_webdriver, retry_count=3):
        kwargs = get_args(driver=splinter_webdriver,
                          download_dir=splinter_file_download_dir,
                          download_ftypes=splinter_download_file_types,
                          firefox_pref=splinter_firefox_profile_preferences,
                          firefox_prof_dir=splinter_firefox_profile_directory,
                          remote_url=splinter_remote_url,
                          executable=splinter_webdriver_executable,
                          headless=splinter_headless,
                          driver_kwargs=splinter_driver_kwargs)
        try:
            return splinter_browser_class(
                splinter_webdriver, visit_condition=splinter_browser_load_condition,
                visit_condition_timeout=splinter_browser_load_timeout,
>               wait_time=splinter_wait_time, **kwargs
            )

/usr/local/lib/python3.6/site-packages/pytest_splinter/plugin.py:517: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = ('chrome',), kwargs = {'headless': True, 'wait_time': 5}
visit_condition = <function splinter_browser_load_condition.<locals>.<lambda> at 0x7ff49aaaf048>
visit_condition_timeout = 10

    def Browser(*args, **kwargs):
        """Emulate splinter's Browser."""
        visit_condition = kwargs.pop('visit_condition')
        visit_condition_timeout = kwargs.pop('visit_condition_timeout')
>       browser = splinter.Browser(*args, **kwargs)

/usr/local/lib/python3.6/site-packages/pytest_splinter/plugin.py:72: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

driver_name = 'chrome', retry_count = 3, args = ()
kwargs = {'headless': True, 'wait_time': 5}
driver = <class 'splinter.driver.webdriver.chrome.WebDriver'>

    def Browser(driver_name="firefox", retry_count=3, *args, **kwargs):
        """
        Returns a driver instance for the given name.
    
        When working with ``firefox``, it's possible to provide a profile name
        and a list of extensions.
    
        If you don't provide any driver_name, then ``firefox`` will be used.
    
        If there is no driver registered with the provided ``driver_name``, this
        function will raise a :class:`splinter.exceptions.DriverNotFoundError`
        exception.
        """
    
        try:
            driver = _DRIVERS[driver_name]
        except KeyError:
            raise DriverNotFoundError("No driver for %s" % driver_name)
    
>       return get_driver(driver, *args, **kwargs)

/usr/local/lib/python3.6/site-packages/splinter/browser.py:90: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

driver = <class 'splinter.driver.webdriver.chrome.WebDriver'>, retry_count = 3
args = (), kwargs = {'headless': True, 'wait_time': 5}, _ = 2

    def get_driver(driver, retry_count=3, *args, **kwargs):
        """Try to instantiate the driver.
    
        Common selenium errors are caught and a retry attempt occurs.
        This can mitigate issues running on Remote WebDriver.
    
        """
        for _ in range(retry_count):
            try:
                return driver(*args, **kwargs)
            except (IOError, HTTPException, WebDriverException, MaxRetryError) as e:
                pass
    
>       raise e
E       UnboundLocalError: local variable 'e' referenced before assignment

/usr/local/lib/python3.6/site-packages/splinter/browser.py:68: UnboundLocalError
```

Dockerfile I am using:

```
FROM python:3.6-slim-buster

# Set working directory
WORKDIR /python/tests/automation/

# Install third-party dependencies
RUN pip install pipenv

COPY Pipfile .
COPY Pipfile.lock .
RUN pipenv install --system --deploy --ignore-pipfile

# Copy automation data and configuration
COPY . .

# Clear .pyc files for conftest.py import
RUN find . -name *.pyc -delete

# Install Google Chrome
RUN apt-get update && \
    apt-get install -y wget dpkg && \
    wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb && \
    dpkg -i google-chrome-stable_current_amd64.deb; apt-get -fy install

RUN webdriverdownloader chrome

RUN pytest -m splinter
```

**Comment by jmt-transloc** [comment]
To add some additional context, this only occurs when using either pytest-splinter or Splinter within a Docker container. 

Running pytest-selenium or Selenium with the above Docker set up does not error out.

Feel free to contact me for any additional information I can provide.

**Comment by jsfehler** [comment]
@jmt-transloc If you switch from the 0.13.0 release of splinter to the master branch, the error you're getting will be raised correctly. This will make debugging this issue easier.

**Comment by jmt-transloc** [comment]
> @jmt-transloc If you switch from the 0.13.0 release of splinter to the master branch, the error you're getting will be raised correctly. This will make debugging this issue easier.

Thanks for your reply! I'll give that a shot and report back.

**Comment by tchabaud** [comment]
Hello,

I'm having troubles trying to run in a Docker container too. I'm using an Alpine 3.11 based container, with chromium and chromedriver installed.

I launch the browser like this :

```python
    opts = Options()
    opts.add_argument("--no-sandbox")
    opts.add_argument("--disable-setuid-sandbox")
    opts.add_argument("--disable-gpu")
    opts.add_argument("--disable-software-rasterizer")
    opts.add_argument('--start-maximized')

    with Browser('chrome', headless=True) as browser:
        browser.driver = webdriver.Chrome('/usr/bin/chromedriver', chrome_options=opts)
        browser.visit('https://google.com')
```

and it always fail with the following error : 

```
selenium.common.exceptions.WebDriverException: Message: unknown error: Chrome failed to start: exited abnormally
  (unknown error: DevToolsActivePort file doesn't exist)
  (The process started from chrome location /usr/lib/chromium/chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.)

```
I'm using splinter Github master version, installed with : 

```sh
python -m pip install --upgrade "git+https://github.com/cobrateam/splinter.git"
```

What am I doing wrong ?

**Comment by tchabaud** [comment]
Ok just find my mistake.

Selenium Chromedriver options was no correctly set, with the following code it works fine : 

```python
import logging

from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from splinter import Browser

logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
                    level=logging.DEBUG)

logger = logging.getLogger(__name__)

opts = Options()
opts.add_argument("--disable-gpu")
opts.add_argument("--no-sandbox")
opts.add_argument("--disable-setuid-sandbox")
opts.add_argument("--disable-software-rasterizer")
opts.add_argument('--start-maximized')

logging.info('Trying to start Chromium via Chromedriver ...')

selenium_opts = {'options': opts}

with Browser('chrome', headless=True, **selenium_opts) as browser:
    browser.visit('https://google.com')

logging.info('Hello Chromium !')
```
Sorry for the noise :disappointed:

**Comment by jmt-transloc** [comment]
Going to close this out as my team and I decided to not move forward with Splinter.
